### ISSUE TITLE
Can't configure amplify with AmplifyAPI and AmplifyAuthCognito

### ISSUE THREAD

**Issue by KyMback** [aws-amplify/amplify-flutter#913]
**Describe the bug**
When i try to configure amplify (via `Amplify.configure(stringConfiguration);`) with rest api and with cognito auth i get the following error:
```
I/amplify:flutter:auth_cognito(22616): Added Auth plugin
I/amplify:flutter:api(22616): Added API plugin
D/AWSMobileClient(22616): Using the SignInProviderConfig from `awsconfiguration.json`.
D/AWSMobileClient(22616): Inspecting user state details
W/CognitoUserSession(22616): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(22616): signalTokensNotAvailable
W/InternalCallback(22616): Library attempted to call user callback twice, expected only once
E/flutter (22616): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: AmplifyException(message: Failed to retrieve auth token from Cognito provider., recoverySuggestion: Check the application logs for details., underlyingException: ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.})
E/flutter (22616): #0      AmplifyClass.configure (package:amplify_flutter/amplify.dart:173:9)
E/flutter (22616): <asynchronous suspension>
E/flutter (22616): #1      _MyHomePageState.configure (package:test_amplify_bug/main.dart:71:5)
E/flutter (22616): <asynchronous suspension>
E/flutter (22616): 
```
I noticed that this behavior reproducible when using amplify libs since 0.2.3 version. I use these libs:

```
  amplify_api: ^0.2.4
  amplify_flutter: ^0.2.4
  amplify_auth_cognito: ^0.2.4
```

```dart
  @override
  void initState() {
    configure();
    super.initState();
  }

  Future configure() async {
    if (Amplify.isConfigured) {
      return Future.value();
    }

    await Amplify.addPlugins([AmplifyAuthCognito(), AmplifyAPI()]);

    final stringConfiguration = AmplifyConfigurationBuilder.build(
      cognitoIdentityPoolId: 'cognitoIdentityPoolId',
      cognitoIdentityRegion: ' cognitoIdentityRegion',
      cognitoUserPoolId: 'cognitoUserPoolId',
      cognitoUserPoolAppClientId: 'cognitoUserPoolAppClientId:',
      cognitoUserPoolAppClientSecret: 'cognitoUserPoolAppClientSecret',
      cognitoUserPoolRegion: 'cognitoUserPoolRegion',
      authenticationFlowType: 'USER_PASSWORD_AUTH',
      apiGatewayEndpoint: 'apiGatewayEndpoint',
      apiGatewayRegion: 'apiGatewayRegion',
      apiName: 'api',
    );
    await Amplify.configure(stringConfiguration);
  }
```
And AmplifyConfigurationBuilder
```dart
class AmplifyConfigurationBuilder {
  static String build({
    required String cognitoIdentityPoolId,
    required String cognitoIdentityRegion,
    required String cognitoUserPoolId,
    required String cognitoUserPoolAppClientId,
    required String cognitoUserPoolAppClientSecret,
    required String cognitoUserPoolRegion,
    required String authenticationFlowType,
    required String apiName,
    required String apiGatewayEndpoint,
    required String apiGatewayRegion,
  }) {
    return '''
{
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "$cognitoIdentityPoolId",
                            "Region": "$cognitoIdentityRegion"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "$cognitoUserPoolId",
                        "AppClientId": "$cognitoUserPoolAppClientId",
                        "AppClientSecret": "$cognitoUserPoolAppClientSecret",
                        "Region": "$cognitoUserPoolRegion"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "$authenticationFlowType"
                    }
                }
            }
        }
    },
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "api": {
                    "endpointType": "REST",
                    "endpoint": "$apiGatewayEndpoint",
                    "region": "$apiGatewayRegion",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    }
}
''';
  }
}
```

**To Reproduce**
1. flutter create myapp
2. Use the above code
3. Run the app

**Expected behavior**
No errors. Possibility to use AmplifyAPI with AmplifyAuthCognito.

**Screenshots**

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[+] Android
[?] iOS

<details>

  <summary>Output of <code>flutter doctor -v</code></summary>
  
  ```
[✓] Flutter (Channel stable, 2.5.1, on CentOS Linux 7 (Core) 3.10.0-1127.el7.x86_64, locale en_US.UTF-8)
    • Flutter version 2.5.1 at /home/user/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (7 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/user/Android/Sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/user/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000
    • Flutter plugin version 60.1.2
    • Dart plugin version 203.8292
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 93.0.4577.82

! Doctor found issues in 1 category.
``

</details>

**Smartphone (please complete the following information):**
 - Device: [Pixel 4 (emulator)]
 - OS: [Android 11 (API 30)]

**Additional context**
Add any other context about the problem here.

**Comment by dnys1** [comment]
Hi @KyMback, thanks for your detailed report. I believe the necessary `authorizationType` for authenticating against an Amplify REST API with Cognito User Pools is `AWS_IAM`. Could you make the change and let me know if that solves your problem?

I'd like to learn more about your use case- have you found the Amplify CLI to not work well for you when it comes to generating configuration files? We don't currently support custom configurations like this, so future issues may be more difficult to troubleshoot.

**Comment by KyMback** [comment]
Thanks for the response.
When i change `authorizationType` to `AWS_IAM` the issue on `configure` is fixed but i still can't authorize my restapi request via cognito authorizer (i get 401 status code).

I don't use amplify cli because i use already existing aws resources.

On the cloud I have REST API with cognito authorizer. When I try to login and send request via postman it works fine, but in mobile no. Also I noticed that when i manually set up `Authorization` (with id token which i receive via `Amplify.Auth.fetchAuthSession`) it also fails.

**Comment by dnys1** [comment]
Interesting, okay, I will need to do some further investigation, then. Thank you for sharing - this may either be a blind spot in our SDK or just an unsupported use case. I'm not sure which, yet, but I will find out and get back to you soon.

**Comment by sanchitd5** [comment]
+1 facing the same issue

**Comment by jasonhtpham** [comment]
More details:

Basically, if we specify `authorizationType` as `AMAZON_COGNITO_USER_POOLS` the api call will work fine, but the authorization will randomly work sometimes (not stable).

When we change the `authorizationType` to `AWS_IAM` the authorization seems to work fine but the api calls will give errors.

**Comment by dnys1** [comment]
@sanchitd5 @jasonhtpham Can you both confirm you are all using a Cognito User Pool and API Gateway configured outside of Amplify, and not one configured through the CLI?

**Comment by dnys1** [comment]
@KyMback @sanchitd5 @jasonhtpham In order to resolve this, I need some more information. 

Please be specific about your API Gateway authorizer configuration:
1. Confirm you are trying to use the `COGNITO_USER_POOLS` authorization type in your gateway.
  a. If not, then the `AMAZON_COGNITO_USER_POOLS` authorization type will not work.
2. If so, are you using Cognito identity tokens or access tokens to restrict access?
  a. If access tokens, have you enabled a custom resource server with appropriate scopes in Cognito? (See [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) and [here](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html))
3. Which OAuth flows are you using in your user pool client?
4. Are you using a Cognito Identity Pool?
  a. If so, does it allow unauthenticated logins? And does it support identity providers besides Cognito?

**Comment by sanchitd5** [comment]
We were using react-native before, we migrated to flutter recently 

1. We are using COGNITO_USER_POOLS authorization type in the gateway
2. Cognito identity tokens
3. Allowed OAuth Flows: 'Authorization code grant' and 'Implicit grant'
4.  not using Cognito Identity Pool

we did created a indentity Id for the flutter app but in the react native app we didn't had to
we are using id tokens, and use cognito user pools without the Cognito Identity Pool

**Comment by iThink32** [comment]
@ragingsquirrel3 what about those using storage? There is no option to manually add headers can you also throw some light on this pls. Ive been stuck and requesting help for a long time now.
https://github.com/aws-amplify/amplify-ios/issues/1439

**Comment by dnys1** [comment]
@iThink32 - this issue seems unrelated to the discussion in this thread. Please see the response from Michael in the iOS thread: https://github.com/aws-amplify/amplify-ios/issues/1439#issuecomment-969117359. Storage requests are signed based off the logged-in user, and use an identity pool for getting AWS credentials, not a user pool. There is no way to manually configure these headers in the native libraries, so this would not be possible to support in Flutter.

Feel free to create another issue in this repo if you'd like to continue the discussion here.

**Comment by KyMback** [comment]
Thank you for the answer. I'm sorry but my back-end engineer can't check your solution anytime soon for some reasons, but i hope it really work...

**Comment by dnys1** [comment]
Closing for now since it is outside the scope of the Amplify CLI and there is a workaround available. Please re-open if new details emerge or if the mentioned workaround does not fix the problem.
